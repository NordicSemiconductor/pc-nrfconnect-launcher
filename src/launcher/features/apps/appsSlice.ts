/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';

import {
    App,
    AppSpec,
    DownloadableApp,
    DownloadableAppInfo,
    isDownloadable,
    isInstalled,
    LaunchableApp,
    LocalApp,
    updateAvailable,
} from '../../../ipc/apps';
import { Progress } from '../../../ipc/downloadProgress';
import { allStandardSourceNames, SourceName } from '../../../ipc/sources';
import type { RootState } from '../../store';
import { getAppsFilter } from '../filter/filterSlice';

type AppProgess = {
    isInstalling: boolean;
    isUpdating: boolean;
    isRemoving: boolean;
    fraction: number;
};

const notInProgress = (): AppProgess => ({
    isInstalling: false,
    isUpdating: false,
    isRemoving: false,
    fraction: 0,
});

const appNotInProgress = <X>(app: X) => ({
    ...app,
    progress: notInProgress(),
});

type DownloadableAppWithProgress = DownloadableApp & {
    progress: AppProgess;
};

export type DisplayedApp = LocalApp | DownloadableAppWithProgress;

export type State = {
    localApps: LocalApp[];
    downloadableApps: DownloadableAppWithProgress[];
    lastUpdateCheckDate?: Date;
    isDownloadingLatestAppInfo: boolean;
    isConfirmLaunchDialogVisible: boolean;
    confirmLaunchText?: string;
    confirmLaunchApp?: LaunchableApp;
};

const initialState: State = {
    localApps: [],
    downloadableApps: [],
    isDownloadingLatestAppInfo: false,
    isConfirmLaunchDialogVisible: false,
};

const equalsSpec = (specOfSoughtApp: AppSpec) => (app: App) =>
    app.source === specOfSoughtApp.source && app.name === specOfSoughtApp.name;

const notEqualsSpec = (specOfSoughtApp: AppSpec) => (app: App) =>
    !equalsSpec(specOfSoughtApp)(app);

const updateApp = <AppType extends App>(
    specOfAppToUpdate: AppSpec,
    apps: AppType[],
    update: (app: AppType) => void
) => {
    const appToUpdate = apps.find(equalsSpec(specOfAppToUpdate));

    if (appToUpdate != null) {
        update(appToUpdate);
    } else {
        console.error(
            `No app ${appToUpdate} found in the existing downloadable apps though there is supposed to be one.`
        );
    }
};

const resetProgress = (
    specOfAppToUpdate: AppSpec,
    apps: DownloadableAppWithProgress[]
) => {
    updateApp(specOfAppToUpdate, apps, app => {
        app.progress = notInProgress();
    });
};

const addUninstalledApp = (
    apps: DownloadableAppWithProgress[],
    updatedAppInfo: DownloadableAppInfo
) => {
    apps.push({
        ...appNotInProgress(updatedAppInfo),
        currentVersion: undefined,
    });
};

const updateInstalledApp = (
    app: DownloadableAppWithProgress,
    updatedAppInfo: DownloadableAppInfo
) => {
    app.latestVersion = updatedAppInfo.latestVersion;
    app.versions = updatedAppInfo.versions;
};

const updateUninstalledApp = (
    app: DownloadableAppWithProgress,
    updatedAppInfo: DownloadableAppInfo
) => {
    Object.assign(app, updatedAppInfo);
};

const mergeDownloadedAppInfoIn = (
    apps: DownloadableAppWithProgress[],
    updatedAppInfos: DownloadableAppInfo[]
) => {
    updatedAppInfos.forEach(updatedAppInfo => {
        const existingApp = apps.find(equalsSpec(updatedAppInfo));

        const appIsStillUnknown = existingApp == null;
        if (appIsStillUnknown) {
            addUninstalledApp(apps, updatedAppInfo);
        } else if (isInstalled(existingApp)) {
            updateInstalledApp(existingApp, updatedAppInfo);
        } else {
            updateUninstalledApp(existingApp, updatedAppInfo);
        }
    });
};

const slice = createSlice({
    name: 'apps',
    initialState,
    reducers: {
        // Local apps
        setAllLocalApps(
            state,
            { payload: localApps }: PayloadAction<LocalApp[]>
        ) {
            state.localApps = [...localApps];
        },

        addLocalApp(state, { payload: newApp }: PayloadAction<LocalApp>) {
            state.localApps.push(newApp);
        },

        removeLocalApp(state, { payload: appName }: PayloadAction<string>) {
            state.localApps = state.localApps.filter(
                app => app.name !== appName
            );
        },

        // Downloadable apps
        addDownloadableApps(
            state,
            { payload: additionalApps }: PayloadAction<DownloadableApp[]>
        ) {
            state.downloadableApps.push(
                ...additionalApps.map(appNotInProgress)
            );
        },

        updateDownloadableAppInfo(
            state,
            { payload: updatedApp }: PayloadAction<DownloadableApp>
        ) {
            state.downloadableApps = state.downloadableApps
                .filter(notEqualsSpec(updatedApp))
                .concat({
                    ...updatedApp,
                    progress: notInProgress(),
                });
        },

        removeAppsOfSource(
            state,
            { payload: sourceName }: PayloadAction<SourceName>
        ) {
            state.downloadableApps = state.downloadableApps.filter(
                app => app.source !== sourceName
            );
        },

        // Update downloadable app infos
        updateDownloadableAppInfosStarted(state) {
            state.isDownloadingLatestAppInfo = true;
        },
        updateDownloadableAppInfos(
            state,
            {
                payload,
            }: PayloadAction<{
                updatedAppInfos: DownloadableAppInfo[];
                updateCheckDate?: Date;
            }>
        ) {
            mergeDownloadedAppInfoIn(
                state.downloadableApps,
                payload.updatedAppInfos
            );

            state.isDownloadingLatestAppInfo = false;
            state.lastUpdateCheckDate = payload.updateCheckDate ?? new Date();
        },
        updateDownloadableAppInfosFailed(state) {
            state.isDownloadingLatestAppInfo = false;
        },

        // Update progress
        updateAppProgress(state, { payload }: PayloadAction<Progress>) {
            updateApp(payload.app, state.downloadableApps, app => {
                app.progress.fraction = payload.progressFraction;
            });
        },

        resetAppProgress(state, { payload: app }: PayloadAction<AppSpec>) {
            resetProgress(app, state.downloadableApps);
        },

        // Install downloadable app
        installDownloadableAppStarted(
            state,
            { payload: appToInstall }: PayloadAction<AppSpec>
        ) {
            updateApp(appToInstall, state.downloadableApps, app => {
                app.progress.isInstalling = true;
            });
        },

        // Update downloadable app
        updateDownloadableAppStarted(
            state,
            { payload: appToUpdate }: PayloadAction<AppSpec>
        ) {
            updateApp(appToUpdate, state.downloadableApps, app => {
                app.progress.isUpdating = true;
            });
        },

        // Remove downloadable app
        removeDownloadableAppStarted(
            state,
            { payload: appToRemove }: PayloadAction<AppSpec>
        ) {
            updateApp(appToRemove, state.downloadableApps, app => {
                app.progress.isRemoving = true;
            });
        },
        removeDownloadableAppSuccess(
            state,
            { payload: removedApp }: PayloadAction<AppSpec>
        ) {
            resetProgress(removedApp, state.downloadableApps);

            updateApp(removedApp, state.downloadableApps, app => {
                app.currentVersion = undefined;
            });
        },

        // Confirm launch dialog
        showConfirmLaunchDialog(
            state,
            { payload }: PayloadAction<{ text: string; app: LaunchableApp }>
        ) {
            state.isConfirmLaunchDialogVisible = true;
            state.confirmLaunchText = payload.text;
            state.confirmLaunchApp = payload.app;
        },
        hideConfirmLaunchDialog(state) {
            state.isConfirmLaunchDialogVisible = false;
            state.confirmLaunchText = initialState.confirmLaunchText;
            state.confirmLaunchApp = initialState.confirmLaunchApp;
        },
    },
});

export default slice.reducer;

export const {
    addDownloadableApps,
    addLocalApp,
    hideConfirmLaunchDialog,
    installDownloadableAppStarted,
    removeAppsOfSource,
    removeDownloadableAppStarted,
    removeDownloadableAppSuccess,
    removeLocalApp,
    resetAppProgress,
    setAllLocalApps,
    showConfirmLaunchDialog,
    updateDownloadableAppInfo,
    updateDownloadableAppInfos,
    updateDownloadableAppInfosFailed,
    updateDownloadableAppInfosStarted,
    updateDownloadableAppStarted,
    updateAppProgress,
} = slice.actions;

export const getAllApps = (state: RootState): DisplayedApp[] => {
    const { downloadableApps, localApps } = state.apps;

    return [...localApps, ...downloadableApps];
};

export const getAllSourceNamesSorted = (state: RootState): SourceName[] => {
    const allSources = [
        ...new Set(getAllApps(state).map(({ source }) => source)),
    ];

    const withoutStandardSources = allSources.filter(
        source => !allStandardSourceNames.includes(source)
    );

    return [
        ...allStandardSourceNames,
        ...withoutStandardSources.sort((a, b) => a.localeCompare(b)),
    ];
};

export const getDownloadableApp =
    ({ source, name }: { name?: string; source?: SourceName }) =>
    (state: RootState) =>
        state.apps.downloadableApps.find(
            x => x.source === source && x.name === name
        );

export const isAppUpdateAvailable = (state: RootState) =>
    state.apps.downloadableApps.find(
        app => isInstalled(app) && updateAvailable(app)
    ) != null;

export const getUpdateCheckStatus = (state: RootState) => ({
    isCheckingForUpdates: state.apps.isDownloadingLatestAppInfo,
    lastUpdateCheckDate: state.apps.lastUpdateCheckDate,
});

export const getUpdatableVisibleApps = (state: RootState) =>
    state.apps.downloadableApps
        .filter(getAppsFilter(state))
        .filter(app => isInstalled(app) && updateAvailable(app));

export const getConfirmLaunch = (state: RootState) => ({
    isDialogVisible: state.apps.isConfirmLaunchDialogVisible,
    text: state.apps.confirmLaunchText,
    app: state.apps.confirmLaunchApp,
});

export const isInProgress = (
    app: DisplayedApp
): app is DownloadableAppWithProgress =>
    isDownloadable(app) &&
    (app.progress.isInstalling ||
        app.progress.isUpdating ||
        app.progress.isRemoving);
