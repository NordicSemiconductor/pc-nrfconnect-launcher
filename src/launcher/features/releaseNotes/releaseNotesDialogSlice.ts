/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';

import type { RootState } from '../..';

export type State = {
    source?: string;
    name?: string;
};

const initialState: State = {};

const slice = createSlice({
    name: 'releaseNotesDialog',
    initialState,
    reducers: {
        show(
            state,
            { payload }: PayloadAction<{ name: string; source: string }>
        ) {
            state.name = payload.name;
            state.source = payload.source;
        },
        hide() {
            return initialState;
        },
    },
});

export default slice.reducer;

export const { show, hide } = slice.actions;

export const getReleaseNotesDialog = (state: RootState) =>
    state.releaseNotesDialog;
