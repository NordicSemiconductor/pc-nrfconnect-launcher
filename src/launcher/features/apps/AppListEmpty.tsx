/*
 * Copyright (c) 2023 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React, {
    MouseEventHandler,
    ReactNode,
    useEffect,
    useState,
} from 'react';
import { ExternalLink } from '@nordicsemiconductor/pc-nrfconnect-shared';

import { useLauncherDispatch, useLauncherSelector } from '../../util/hooks';
import { startUpdateProcess } from '../process/updateProcess';
import { getShouldCheckForUpdatesAtStartup } from '../settings/settingsSlice';
import { getNoAppsExist } from './appsSlice';

const Box = ({ children }: { children: ReactNode }) => (
    <div className="tw-preflight tw-grid tw-flex-1 tw-place-items-center">
        <div className="tw-text-center">{children}</div>
    </div>
);

const InlineButton = ({
    onClick,
    children,
}: {
    children: ReactNode;
    onClick: MouseEventHandler;
}) => (
    <button
        className="tw-inline tw-border-0 tw-p-0 tw-text-primary"
        type="button"
        onClick={onClick}
    >
        {children}
    </button>
);

const CheckForUpdatesDisabled = () => {
    const dispatch = useLauncherDispatch();

    return (
        <Box>
            <p>No apps are loaded from the server yet.</p>
            <p>
                You have “Check for updates at startup” disabled in the
                settings.
            </p>
            <p>
                You can enable it there or now just{' '}
                <InlineButton
                    onClick={() => dispatch(startUpdateProcess(true))}
                >
                    check once for updates
                </InlineButton>
                .
            </p>
        </Box>
    );
};

const NotLoadedYet = () => {
    const [justStarted, setJustStarted] = useState(true);

    useEffect(() => {
        setTimeout(() => setJustStarted(false), 2000);
    }, []);

    return justStarted ? null : (
        <Box>
            <p>
                The list of apps is not yet loaded from{' '}
                <ExternalLink href="https://files.nordicsemi.com" />.
            </p>
            <p>Make sure you can reach that server.</p>
        </Box>
    );
};

const NoApps = () => {
    const shouldCheckForUpdatesAtStartup = useLauncherSelector(
        getShouldCheckForUpdatesAtStartup,
    );

    return shouldCheckForUpdatesAtStartup ? (
        <NotLoadedYet />
    ) : (
        <CheckForUpdatesDisabled />
    );
};

const AllFilteredOut = () => (
    <Box>
        <p>No apps shown because of the selected filters.</p>
        <p>Change those to display apps again.</p>
    </Box>
);

export default () => {
    const noAppsExist = useLauncherSelector(getNoAppsExist);

    return noAppsExist ? <NoApps /> : <AllFilteredOut />;
};
