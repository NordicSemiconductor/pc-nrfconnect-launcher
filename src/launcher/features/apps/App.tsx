/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';
import Button from 'react-bootstrap/Button';
import ButtonToolbar from 'react-bootstrap/ButtonToolbar';
import Col from 'react-bootstrap/Col';
import Dropdown from 'react-bootstrap/Dropdown';
import DropdownButton from 'react-bootstrap/DropdownButton';
import ListGroup from 'react-bootstrap/ListGroup';
import ProgressBar from 'react-bootstrap/ProgressBar';
import Row from 'react-bootstrap/Row';
import { openUrl } from 'pc-nrfconnect-shared';

import { DownloadableApp, LocalApp } from '../../../ipc/apps';
import { createDesktopShortcut } from '../../../ipc/createDesktopShortcut';
import { useLauncherDispatch, useLauncherSelector } from '../../util/hooks';
import { show as showReleaseNotes } from '../releaseNotes/releaseNotesDialogSlice';
import AppIcon from './AppIcon';
import {
    checkEngineAndLaunch,
    installDownloadableApp,
    removeDownloadableApp,
} from './appsEffects';
import {
    getIsAppInstalling,
    getIsAppRemoving,
    getIsAppUpgrading,
} from './appsSlice';

import '../../../../resources/css/launcher.scss';

const AppItem: React.FC<{
    app: LocalApp | (DownloadableApp & { progress?: number });
    isDisabled: boolean;
}> = ({ app, isDisabled }) => {
    const dispatch = useLauncherDispatch();

    const isInstalling = useLauncherSelector(getIsAppInstalling(app));
    const isUpgrading = useLauncherSelector(getIsAppUpgrading(app));
    const isRemoving = useLauncherSelector(getIsAppRemoving(app));

    return (
        <ListGroup.Item>
            <Row noGutters className="py-1">
                <Col xs="auto" className="d-flex align-items-start my-2 mr-3">
                    <AppIcon app={app} />
                </Col>
                <Col>
                    <div className="h8">{app.displayName || app.name}</div>
                    <div className="small text-muted">{app.description}</div>
                    <div className="small text-muted-more">
                        {app.source}
                        {app.isInstalled && <>, v{app.currentVersion}</>}
                        {app.isInstalled &&
                            app.isDownloadable &&
                            app.upgradeAvailable && (
                                <> (v{app.latestVersion} available)</>
                            )}
                        {!app.isInstalled && app.latestVersion && (
                            <>, v{app.latestVersion}</>
                        )}
                    </div>
                </Col>
                <Col
                    xs="auto"
                    className="d-flex align-items-center my-3 pl-3 ml-auto"
                >
                    <ButtonToolbar className="wide-btns">
                        {app.isInstalled &&
                            app.isDownloadable &&
                            app.upgradeAvailable && (
                                <Button
                                    variant="outline-primary"
                                    title={`Update ${app.displayName}`}
                                    disabled={isDisabled}
                                    onClick={() =>
                                        dispatch(showReleaseNotes(app))
                                    }
                                >
                                    {isUpgrading ? 'Updating...' : 'Update'}
                                </Button>
                            )}
                        {app.isInstalled && (
                            <Button
                                title={`Open ${app.displayName}`}
                                disabled={isDisabled}
                                onClick={() =>
                                    dispatch(checkEngineAndLaunch(app))
                                }
                            >
                                Open
                            </Button>
                        )}
                        {!app.isInstalled && (
                            <Button
                                variant="outline-secondary"
                                title={`Install ${app.displayName}`}
                                disabled={isDisabled}
                                onClick={() =>
                                    dispatch(installDownloadableApp(app))
                                }
                            >
                                {isInstalling ? 'Installing...' : 'Install'}
                            </Button>
                        )}
                        <DropdownButton
                            variant={
                                app.isInstalled
                                    ? 'outline-primary'
                                    : 'outline-secondary'
                            }
                            title=""
                            alignRight
                        >
                            {app.isDownloadable && app.homepage != null && (
                                <Dropdown.Item
                                    title="Go to app website"
                                    onClick={() => openUrl(app.homepage!)} // eslint-disable-line @typescript-eslint/no-non-null-assertion
                                >
                                    More info
                                </Dropdown.Item>
                            )}
                            {app.isDownloadable && app.releaseNote && (
                                <Dropdown.Item
                                    title="Show release notes"
                                    onClick={() =>
                                        dispatch(showReleaseNotes(app))
                                    }
                                >
                                    Release notes
                                </Dropdown.Item>
                            )}
                            {app.isInstalled && (
                                <Dropdown.Item
                                    title="Create a desktop shortcut for this app"
                                    onClick={() => createDesktopShortcut(app)}
                                >
                                    Create shortcut
                                </Dropdown.Item>
                            )}
                            {app.isInstalled && app.isDownloadable && (
                                <Dropdown.Item
                                    title={`Remove ${app.displayName}`}
                                    disabled={isDisabled}
                                    onClick={() =>
                                        dispatch(removeDownloadableApp(app))
                                    }
                                >
                                    {isRemoving
                                        ? 'Uninstalling...'
                                        : 'Uninstall'}
                                </Dropdown.Item>
                            )}
                        </DropdownButton>
                    </ButtonToolbar>
                </Col>
            </Row>
            {app.isDownloadable && app.progress != null && (
                <ProgressBar now={app.progress} />
            )}
        </ListGroup.Item>
    );
};

export default AppItem;
