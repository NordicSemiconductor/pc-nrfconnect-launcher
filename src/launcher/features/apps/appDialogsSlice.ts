/*
 * Copyright (c) 2023 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';

import { DownloadableApp, LaunchableApp } from '../../../ipc/apps';
import type { RootState } from '../../store';

type HiddenDialog = typeof hiddenDialog;
const hiddenDialog = {
    isVisible: false,
} as const;

type VisibleConfirmLaunchDialog = ReturnType<typeof visibleConfirmLaunchDialog>;
const visibleConfirmLaunchDialog = (
    text: React.ReactNode,
    app: LaunchableApp
) => ({ isVisible: true, text, app } as const);

type VisibleInstallOtherVersionDialog = ReturnType<
    typeof visibleInstallOtherVersionDialog
>;
const visibleInstallOtherVersionDialog = (app: DownloadableApp) =>
    ({ isVisible: true, app } as const);

export type State = {
    confirmLaunch: HiddenDialog | VisibleConfirmLaunchDialog;
    installOtherVersion: HiddenDialog | VisibleInstallOtherVersionDialog;
};

const initialState: State = {
    confirmLaunch: hiddenDialog,
    installOtherVersion: hiddenDialog,
};

const slice = createSlice({
    name: 'apps',
    initialState,
    reducers: {
        showConfirmLaunchDialog(
            state,
            {
                payload,
            }: PayloadAction<{
                text: React.ReactNode;
                app: LaunchableApp;
            }>
        ) {
            state.confirmLaunch = visibleConfirmLaunchDialog(
                payload.text,
                payload.app
            );
        },
        hideConfirmLaunchDialog(state) {
            state.confirmLaunch = hiddenDialog;
        },
        showInstallOtherVersionDialog(
            state,
            { payload: app }: PayloadAction<DownloadableApp>
        ) {
            state.installOtherVersion = visibleInstallOtherVersionDialog(app);
        },
        hideInstallOtherVersionDialog(state) {
            state.installOtherVersion = hiddenDialog;
        },
    },
});

export default slice.reducer;

export const {
    hideConfirmLaunchDialog,
    hideInstallOtherVersionDialog,
    showConfirmLaunchDialog,
    showInstallOtherVersionDialog,
} = slice.actions;

export const getConfirmLaunchDialog = (state: RootState) =>
    state.appDialogs.confirmLaunch;

export const getInstallOtherVersionDialog = (state: RootState) =>
    state.appDialogs.installOtherVersion;
