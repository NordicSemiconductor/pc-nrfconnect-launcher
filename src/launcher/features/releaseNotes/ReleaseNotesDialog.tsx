/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';
import Button from 'react-bootstrap/Button';
import Modal from 'react-bootstrap/Modal';
import ReactMarkdown from 'react-markdown';

import { InstalledDownloadableApp } from '../../../ipc/apps';
import { useLauncherDispatch, useLauncherSelector } from '../../util/hooks';
import { upgradeDownloadableApp } from '../apps/appsEffects';
import { getDownloadableApp } from '../apps/appsSlice';
import { getReleaseNotesDialog, hide } from './releaseNotesDialogSlice';

type Uninstalled = { isInstalled: false };
const canBeInstalledOrUpdated = (
    app?: Uninstalled | InstalledDownloadableApp
) => !app?.isInstalled || app?.currentVersion !== app?.latestVersion;

export default () => {
    const dispatch = useLauncherDispatch();
    const appToDisplay = useLauncherSelector(getReleaseNotesDialog);
    const app = useLauncherSelector(getDownloadableApp(appToDisplay)) ?? {
        name: appToDisplay.name ?? '',
        source: appToDisplay.source ?? '',
        description: '',
        url: '',
        isInstalled: false,
        displayName: '',
        releaseNote: undefined,
        latestVersion: '',
    };

    const isVisible = appToDisplay.name != null;

    const hideDialog = () => dispatch(hide());

    return (
        <Modal
            show={isVisible}
            onHide={hideDialog}
            size="xl"
            scrollable
            key="releaseNotes"
        >
            <Modal.Header>
                <Modal.Title>Release notes for {app.displayName}</Modal.Title>
            </Modal.Header>
            <Modal.Body className="release-notes">
                <ReactMarkdown
                    source={app.releaseNote ?? ''}
                    linkTarget="_blank"
                />
            </Modal.Body>
            <Modal.Footer>
                {canBeInstalledOrUpdated(app) && (
                    <Button
                        variant="primary"
                        onClick={() => {
                            dispatch(
                                upgradeDownloadableApp(app, app.latestVersion)
                            );
                            hideDialog();
                        }}
                    >
                        {app.isInstalled
                            ? 'Update to latest version'
                            : 'Install'}
                    </Button>
                )}
                <Button variant="outline-primary" onClick={hideDialog}>
                    Close
                </Button>
            </Modal.Footer>
        </Modal>
    );
};
