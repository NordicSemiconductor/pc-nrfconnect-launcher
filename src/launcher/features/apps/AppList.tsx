/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';

import { App as AppType, isInstalled } from '../../../ipc/apps';
import { useLauncherSelector } from '../../util/hooks';
import WithScrollbarContainer from '../../util/WithScrollbarContainer';
import AppFilterBar from '../filter/AppFilterBar';
import { getAppsFilter } from '../filter/filterSlice';
import ReleaseNotesDialog from '../releaseNotes/ReleaseNotesDialog';
import App from './App/App';
import { getAllApps } from './appsSlice';
import ConfirmLaunchDialog from './ConfirmLaunchDialog';
import InstallOtherVersionDialog from './InstallOtherVersionDialog';

const sortByStateAndName = (appA: AppType, appB: AppType) => {
    const cmpInstalled = Number(isInstalled(appB)) - Number(isInstalled(appA));

    const aName = appA.displayName || appA.name;
    const bName = appB.displayName || appB.name;

    return cmpInstalled || aName.localeCompare(bName);
};

export default () => {
    const allApps = useLauncherSelector(getAllApps);
    const appsFilter = useLauncherSelector(getAppsFilter);

    const apps = allApps.filter(appsFilter).sort(sortByStateAndName);

    return (
        <>
            <AppFilterBar />
            <WithScrollbarContainer hasFilter>
                {apps.map(app => (
                    <App key={`${app.name}-${app.source}`} app={app} />
                ))}
            </WithScrollbarContainer>

            <ConfirmLaunchDialog />
            <InstallOtherVersionDialog />
            <ReleaseNotesDialog />
        </>
    );
};
