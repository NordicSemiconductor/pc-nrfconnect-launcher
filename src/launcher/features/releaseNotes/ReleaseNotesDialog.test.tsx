/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';

import type {
    InstalledDownloadableApp,
    UninstalledDownloadableApp,
} from '../../../ipc/apps';
import render from '../../../testrenderer';
import { loadDownloadableAppsSuccess } from '../apps/appsSlice';
import ReleaseNotesDialog from './ReleaseNotesDialog';
import { hide, show } from './releaseNotesDialogSlice';

const unimportantAppProperties = {
    description: 'the test app',
    isDownloadable: true,
    source: 'test source',
    url: 'test url',
    path: '',
    iconPath: '',
    shortcutIconPath: '',
} as const;

const uninstalledApp: UninstalledDownloadableApp = {
    ...unimportantAppProperties,
    name: 'uninstalled',
    displayName: 'uninstalled test app',
    latestVersion: '2.0.0',
    releaseNote: 'Release notes for uninstalled',
    isInstalled: false,
    currentVersion: null,
};
const installedApp: InstalledDownloadableApp = {
    ...unimportantAppProperties,
    name: 'installed',
    displayName: 'installed test app',
    latestVersion: '2.0.0',
    currentVersion: '2.0.0',
    releaseNote: 'Release notes for installed',
    isInstalled: true,
};
const updatableApp: InstalledDownloadableApp = {
    ...unimportantAppProperties,
    name: 'updatable',
    displayName: 'updatable test app',
    latestVersion: '2.0.0',
    currentVersion: '2.1.0',
    releaseNote: 'Release notes for updatable',
    isInstalled: true,
};

const downloadableApps = [uninstalledApp, installedApp, updatableApp];

describe('ReleaseNotesDialog', () => {
    it('is initially invisible', () => {
        expect(
            render(<ReleaseNotesDialog />, [
                loadDownloadableAppsSuccess({ downloadableApps }),
            ]).baseElement
        ).toMatchSnapshot();
    });

    it('is displayed for an uninstalled app', () => {
        expect(
            render(<ReleaseNotesDialog />, [
                loadDownloadableAppsSuccess({ downloadableApps }),
                show(uninstalledApp),
            ]).baseElement
        ).toMatchSnapshot();
    });

    it('is displayed for an up-to-date installed app', () => {
        expect(
            render(<ReleaseNotesDialog />, [
                loadDownloadableAppsSuccess({ downloadableApps }),
                show(installedApp),
            ]).baseElement
        ).toMatchSnapshot();
    });

    it('is displayed for an updatable app', () => {
        expect(
            render(<ReleaseNotesDialog />, [
                loadDownloadableAppsSuccess({ downloadableApps }),
                show(updatableApp),
            ]).baseElement
        ).toMatchSnapshot();
    });

    it('is hidden again after closing the dialog', () => {
        expect(
            render(<ReleaseNotesDialog />, [
                loadDownloadableAppsSuccess({ downloadableApps }),
                show(uninstalledApp),
                hide(),
            ]).baseElement
        ).toMatchSnapshot();
    });
});
