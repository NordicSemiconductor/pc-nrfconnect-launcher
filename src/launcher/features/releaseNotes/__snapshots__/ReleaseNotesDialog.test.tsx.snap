// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReleaseNotesDialog is displayed for an uninstalled app 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade modal show"
    data-focus-visible-added=""
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-xl modal-dialog-scrollable"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div
            class="modal-title h4"
          >
            Release notes for 
            Uninstalled One Test App
          </div>
        </div>
        <div
          class="release-notes modal-body"
        >
          <p>
            Release notes for one uninstalled app
          </p>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Install
          </button>
          <button
            class="btn btn-outline-primary"
            type="button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ReleaseNotesDialog is displayed for an up-to-date installed app 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade modal show"
    data-focus-visible-added=""
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-xl modal-dialog-scrollable"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div
            class="modal-title h4"
          >
            Release notes for 
            Downloadable Another Test App
          </div>
        </div>
        <div
          class="release-notes modal-body"
        >
          <p>
            Release notes for another installed app
          </p>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="btn btn-outline-primary"
            type="button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ReleaseNotesDialog is displayed for an updatable app 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade modal show"
    data-focus-visible-added=""
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-xl modal-dialog-scrollable"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div
            class="modal-title h4"
          >
            Release notes for 
            Downloadable Third Test App
          </div>
        </div>
        <div
          class="release-notes modal-body"
        >
          <p>
            Release notes for a third updatable app
          </p>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Update to latest version
          </button>
          <button
            class="btn btn-outline-primary"
            type="button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ReleaseNotesDialog is hidden again after closing the dialog 1`] = `
<body
  class=""
  style="padding-right: 0px;"
>
  <div />
</body>
`;

exports[`ReleaseNotesDialog is initially invisible 1`] = `
<body>
  <div />
</body>
`;
